<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Demultiplex &mdash; Ontology 0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../_static/documentation_options.js?v=2709fde1"></script>
        <script src="../../_static/doctools.js?v=9a2dae69"></script>
        <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="../../_static/toggleprompt.js?v=5801b3bb"></script>
        <script>let toggleHintShow = 'Click to show';</script>
        <script>let toggleHintHide = 'Click to hide';</script>
        <script>let toggleOpenOnPrint = 'true';</script>
        <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
        <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
        <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Primer Trim" href="primer_trim.html" />
    <link rel="prev" title="Size Filter" href="size_filter.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            Ontology
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../data.html">Example data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../home.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../prepare.html">Prepare</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../implement.html">Implement</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../analyze.html">Analyze</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../upload_data.html">Upload Data</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../status_report.html">Status Report</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="size_filter.html">Size Filter</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Demultiplex</a></li>
<li class="toctree-l3"><a class="reference internal" href="primer_trim.html">Primer Trim</a></li>
<li class="toctree-l3"><a class="reference internal" href="primary_clustering.html">Primary Clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="secondary_clustering.html">Secondary Clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="identification.html">Identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="contig_selection.html">Contig Selection</a></li>
<li class="toctree-l3"><a class="reference internal" href="report.html">Report</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../key_metrics.html">Key Metrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../other_metrics.html">Other Metrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../otu_summary.html">OTU Summary</a></li>
<li class="toctree-l2"><a class="reference internal" href="../upload_to_bold.html">Upload to BOLD</a></li>
<li class="toctree-l2"><a class="reference internal" href="../confirmation.html">Confirmation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../contact_us.html">Contact Us</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Ontology</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../analyze.html">Analyze</a></li>
          <li class="breadcrumb-item"><a href="../status_report.html">Status Report</a></li>
      <li class="breadcrumb-item active">Demultiplex</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/analyze/status_report/demultiplex.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="demultiplex">
<span id="id1"></span><h1>Demultiplex<a class="headerlink" href="#demultiplex" title="Link to this heading"></a></h1>
<p>This section describes the steps involved in demultiplexing reads based on specific Universal Molecular Identifiers (UMIs) at the 5’ end of sequences.</p>
<section id="task-overview">
<h2>Task Overview<a class="headerlink" href="#task-overview" title="Link to this heading"></a></h2>
<p>Demultiplexing allows multiple samples to be sequenced in a single run, significantly reducing costs and maximizing resource efficiency.</p>
<p>The demultiplexing process separates sequences based on UMIs found at both ends of the reads. It then filters reads, moves files into appropriate directories, renames the files based on mapping, and performs additional filtering by read length. The process ends by generating a summary of the read count.</p>
</section>
<section id="steps">
<h2>Steps<a class="headerlink" href="#steps" title="Link to this heading"></a></h2>
<ol class="arabic">
<li><p><strong>Search for UMIs at the 5’ End</strong>:</p>
<p>Using <cite>cutadapt</cite>, the script searches for all 96 UMIs at the 5’ end of sequences. The UMIs are provided through a reference FASTA file, and sequences are demultiplexed into separate files based on their UMI.</p>
<p>Example Code:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cutadapt_trim5_cmd</span> <span class="o">=</span> <span class="p">[</span>
    <span class="nb">str</span><span class="p">(</span><span class="n">cutadapt_path</span><span class="p">),</span>
    <span class="s1">&#39;-j&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cores</span><span class="p">),</span>  <span class="c1"># Number of cores</span>
    <span class="s1">&#39;-e&#39;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">,</span>  <span class="c1"># Maximum error rate</span>
    <span class="s1">&#39;-O&#39;</span><span class="p">,</span> <span class="s1">&#39;12&#39;</span><span class="p">,</span>  <span class="c1"># Minimum overlap</span>
    <span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;file:</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">app_dir</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;data/PRIMERS/PacBio_V2_UMI1-96.fasta&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>  <span class="c1"># Adapter file</span>
    <span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="s1">&#39;trimmed-</span><span class="si">{name}</span><span class="s1">.for5.fasta&#39;</span><span class="p">,</span>  <span class="c1"># Output file pattern</span>
    <span class="s1">&#39;all.fasta&#39;</span>  <span class="c1"># Input file</span>
<span class="p">]</span>
<span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cutadapt_trim5_cmd</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cutadapt<span class="w"> </span>-j<span class="w"> </span>&lt;max_cores&gt;<span class="w"> </span>-e<span class="w"> </span><span class="m">2</span><span class="w"> </span>-O<span class="w"> </span><span class="m">12</span><span class="w"> </span>-g<span class="w"> </span>file:&lt;UMI_file_path&gt;<span class="w"> </span>-o<span class="w"> </span>trimmed-<span class="o">{</span>name<span class="o">}</span>.for5.fasta<span class="w"> </span>all.fasta
</pre></div>
</div>
<p>Where:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">-j</span> <span class="pre">&lt;max_cores&gt;</span></code>: Specifies the number of CPU cores to use.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-e</span> <span class="pre">2</span></code>: Allows a maximum error rate of 2 mismatches.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-O</span> <span class="pre">12</span></code>: Requires a minimum overlap of 12 bases for the adapter (UMI) to match.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-g</span> <span class="pre">file:&lt;UMI_file_path&gt;</span></code>: Specifies the file containing the UMI sequences.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-o</span> <span class="pre">trimmed-{name}.for5.fasta</span></code>: Output file pattern for the demultiplexed reads.</p></li>
</ul>
</div></blockquote>
</li>
<li><p><strong>Reverse Complement Reads</strong>:</p>
<p>For reads that do not contain a 5’ UMI, the script generates their reverse complement using <cite>seqtk</cite>. These reverse complemented reads are then subjected to another round of UMI search at the new 5’ end.</p>
<p>Example Code:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">seqtk_rc_cmd</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">seqtk_path</span><span class="p">),</span> <span class="s1">&#39;seq&#39;</span><span class="p">,</span> <span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="s1">&#39;trimmed-unknown.for5.fasta&#39;</span><span class="p">]</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;trimmed-unknown.for5.rc.fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">seqtk_rc_cmd</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">outfile</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>seqtk<span class="w"> </span>seq<span class="w"> </span>-r<span class="w"> </span>trimmed-unknown.for5.fasta<span class="w"> </span>&gt;<span class="w"> </span>trimmed-unknown.for5.rc.fasta
</pre></div>
</div>
</li>
<li><p><strong>Merge Forward and Reverse Reads</strong>:</p>
<p>The forward and reverse demultiplexed sequences are merged into a single file for each UMI, combining sequences from both orientations into a final set of FASTA files.</p>
<p>Example Code:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1001</span><span class="p">,</span> <span class="mi">1097</span><span class="p">):</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;trimmed-bc</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">.for5.fasta&quot;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_forward</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;trimmed-bc</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">.for3.fasta&quot;</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_reverse</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;bc</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">.fasta&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_merged</span><span class="p">:</span>
        <span class="n">f_merged</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">f_forward</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>  <span class="c1"># Write contents of the forward file</span>
        <span class="n">f_merged</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">f_reverse</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>  <span class="c1"># Write contents of the reverse file</span>
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cat<span class="w"> </span>trimmed-bc&lt;i&gt;.for5.fasta<span class="w"> </span>trimmed-bc&lt;i&gt;.for3.fasta<span class="w"> </span>&gt;<span class="w"> </span>bc&lt;i&gt;.fasta
</pre></div>
</div>
<p>Where <code class="docutils literal notranslate"><span class="pre">&lt;i&gt;</span></code> represents the UMI index.</p>
</li>
<li><p><strong>Clean Up Intermediate Files</strong>:</p>
<p>After the merging process, any temporary or unknown sequence files are removed.</p>
</li>
<li><p><strong>Map and Rename Files</strong>:</p>
<p>Using a mapping file, the script renames the demultiplexed files to meaningful names based on sample IDs. The mapping file contains UMI combinations and their corresponding sample identifiers. The script processes each line of the mapping file, renaming the files accordingly.</p>
<p>Example Code:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mapfile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="nb">next</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>  <span class="c1"># Skip the first line</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">fwd</span><span class="p">,</span> <span class="n">rev</span><span class="p">,</span> <span class="n">sampleid</span><span class="p">,</span> <span class="n">plateid</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[:</span><span class="mi">4</span><span class="p">]</span>
        <span class="n">old_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">fwd</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">rev</span><span class="si">}</span><span class="s2">.fasta&quot;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">new_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sampleid</span><span class="si">}</span><span class="s2">.fasta&quot;</span>
        <span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">old_name</span><span class="p">,</span> <span class="n">new_name</span><span class="p">)</span>
</pre></div>
</div>
</li>
<li><p><strong>Filter by Read Length</strong>:</p>
<p>After demultiplexing, the script filters the reads by length using a specified minimum and maximum threshold.</p>
</li>
<li><p><strong>Generate Read Count Summary</strong>:</p>
<p>A read count is calculated after demultiplexing and filtering. The count is logged and saved in a file named <code class="docutils literal notranslate"><span class="pre">readcounts.txt</span></code>, providing a summary of the total number of reads remaining after the demultiplexing and filtering steps.</p>
</li>
</ol>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="size_filter.html" class="btn btn-neutral float-left" title="Size Filter" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="primer_trim.html" class="btn btn-neutral float-right" title="Primer Trim" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Center for Biodiversity Genomics.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>